src/functions.rs

```diff
@@ -1,69 +1,79 @@
-use crate::error::{Error, GResult, PredicateType};
+use crate::error::{Error, GResult};
 use crate::geometry::Geometry;
 use crate::{AsRawMut, ContextHandle, Geom};
 use geos_sys::*;
 use std::convert::TryFrom;
 use std::ffi::CStr;
 use std::os::raw::c_char;
-use std::str;
+use std::ptr::NonNull;
 
 // We need to cleanup only the char* from geos, the const char* are not to be freed.
 // this has to be checked method by method in geos
 // so we provide 2 method to wrap a char* to a string, one that manage (and thus free) the underlying char*
 // and one that does not free it
-pub(crate) unsafe fn unmanaged_string(raw_ptr: *const c_char, caller: &str) -> GResult<String> {
-    if raw_ptr.is_null() {
-        return Err(Error::NoConstructionFromNullPtr(format!(
-            "{caller}::unmanaged_string",
-        )));
-    }
-    let c_str = CStr::from_ptr(raw_ptr);
-    match str::from_utf8(c_str.to_bytes()) {
-        Ok(s) => Ok(s.to_string()),
-        Err(e) => Err(Error::GenericError(format!(
-            "{caller}::unmanaged_string failed: {e}",
-        ))),
-    }
+pub(crate) unsafe fn unmanaged_string(ptr: NonNull<c_char>) -> GResult<String> {
+    let c_str = CStr::from_ptr(ptr.as_ptr());
+    String::from_utf8(c_str.to_bytes().to_vec())
+        .map_err(|e| Error::GenericError(format!("unmanaged_string failed with {e}")))
 }
 
-pub(crate) unsafe fn managed_string(
-    raw_ptr: *mut c_char,
-    ctx: &ContextHandle,
-    caller: &str,
-) -> GResult<String> {
-    if raw_ptr.is_null() {
-        return Err(Error::NoConstructionFromNullPtr(format!(
-            "{caller}::managed_string",
-        )));
-    }
-    let s = unmanaged_string(raw_ptr, caller);
-    GEOSFree_r(ctx.as_raw(), raw_ptr as *mut _);
+pub(crate) unsafe fn managed_string(ptr: NonNull<c_char>, ctx: &ContextHandle) -> GResult<String> {
+    let s = unmanaged_string(ptr);
+    GEOSFree_r(ctx.as_raw(), ptr.as_ptr().cast());
     s
 }
 
 pub fn version() -> GResult<String> {
-    unsafe { unmanaged_string(GEOSversion(), "version") }
+    unsafe {
+        let Some(v) = NonNull::new(GEOSversion().cast_mut()) else {
+            return Err(Error::GeosError(("GEOSversion", None)));
+        };
+        unmanaged_string(v)
+    }
 }
 
-pub(crate) fn check_geos_predicate(val: i8, p: PredicateType) -> GResult<bool> {
-    match val {
-        1 => Ok(true),
-        0 => Ok(false),
-        _ => Err(Error::GeosFunctionError(p, val as _)),
-    }
+macro_rules! nullcheck {
+    ($func:ident($ctx:ident.as_raw() $(, $($args:expr),* $(,)?)?)) => {{
+        let result = $func($ctx.as_raw()$(, $($args),*)?);
+        std::ptr::NonNull::new(result as *mut _).ok_or_else(|| {
+            $crate::Error::GeosError((stringify!($func), $ctx.get_last_error()))
+        })
+    }};
 }
 
-pub(crate) fn check_ret(val: i32, p: PredicateType) -> GResult<()> {
-    match val {
-        1 => Ok(()),
-        _ => Err(Error::GeosFunctionError(p, val)),
-    }
+macro_rules! errcheck {
```

Comment:
Generally speaking, it's better to avoid macros if possible. In this case, you can have two versions of `errcheck`: one where you pass `errval` and the function returned value and the other where you call the first version with `0`.

Then you have one less macro. :)